### ISSUE TITLE
[diffusiongemma-26B-A4B-it] softmax over a sharded contracting dim produces ±Inf in the following matmul

### ISSUE THREAD

**Issue by ashokkumarkannan1** [tenstorrent/tt-xla#6075]
### Summary

When a `softmax`'s reduction axis is **sharded**, the cross-device combine of the per-shard **maximum** is emitted as an **add** instead of a **maximum**. The softmax shift becomes `max₀+…+max₇` instead of `max(max₀…max₇)`, `exp()` underflows to zero everywhere, and the divide becomes `0/0` → **±Inf**.

Found on DiffusionGemma-26B-A4B-it (`modeling_diffusion_gemma.py:1249`), where `embed_tokens.weight` is vocab-sharded, so the matmul's contracting dim — and hence the softmax's reduced axis — is sharded.

### Root cause (confirmed in the compiled IR)

```mlir
%3 = stablehlo.reduce(%2 init: -inf) applies stablehlo.maximum across dim [2]   // per-shard MAX
%4 = "stablehlo.all_reduce"(%3) ({ stablehlo.add })                             // <-- SUMS the 8 maxima
%6 = stablehlo.subtract %2, broadcast(%4)                                       // shift ~8x too large
%7 = stablehlo.exponential %6                                                   // underflows to 0
%9 = "stablehlo.all_reduce"(%8) ({ stablehlo.add })                             // correct (sum of exp)
%11 = stablehlo.divide %7, broadcast(%9)                                        // 0/0 -> ±Inf
```

`lib/Dialect/StableHLO/Transforms/ShardyCCLToStableHLOCCLPatterns.cpp:310-314` hardcodes an `AddOp` body for **every** `sdy.all_reduce`:

```cpp
// Default the reduction type to sum since shardy does not have
// support for custom reduction types.
addReductionBlock<mlir::stablehlo::AllReduceOp, mlir::stablehlo::AddOp>(...)
```

Correct for the exp-sum and the dot_general partial sums; wrong for the max.

### Repro

`repro_softmax_sharded_matmul_inf.py` — ~30 lines, random data, no model.

```python
V, H, T = 262144, 2816, 256
x = (torch.rand(1, T, V) * 62.0 - 24.75).to(torch.bfloat16)
W = (torch.randn(V, H) * 0.02).to(torch.bfloat16)
fn = lambda a, b: torch.matmul(a.softmax(-1, dtype=torch.float32).to(b.dtype), b)
w = W.to(xla); xs.mark_sharding(w, mesh, ("model", None))   # shard the CONTRACTING dim
out = torch.compile(fn, backend="tt")(x.to(xla), w)
```

### Isolation — it is the *reduced axis* being sharded, not sharding or device count

| configuration | inf | PCC vs CPU |
|---|---|---|
| matmul alone, sharded contracting dim (host softmax) | 0 | 0.999993 |
| fused, **replicated** weight | 0 | 0.999808 |
| fused, sharded on **non-contracting** dim | 0 | 0.999814 |
| fused, sharded on **contracting** dim | **308,128** | nan |

Also: `softmax` alone is clean at V = 4096 … 262144; reproduces at 4-way and 8-way sharding (321,792 / 308,128 inf); **fp32 does not help** (underflows to exactly 0 too); `max|out| = 3.3895e+38` = the bf16 ceiling, i.e. saturation, against a true result of ~8e-2.

### Impact

Self-conditioning runs on 47 of 48 denoising steps. Per-layer decoder PCC, identical captured inputs:

| step | self-conditioning | input embeds | final logits |
|---|---|---|---|
| 1 | off (signal is zero) | 1.000000 | **0.962782** |
| 2 | on (overflows) | 0.359807 | **0.716694** |

The infinities hit the self-conditioning RMSNorm immediately, which normalises them toward zero — so there is no crash and no NaN, just a quietly degraded result. That is why this surfaced as a mediocre PCC rather than an obvious failure.

### Suggested fix

1. **Correct fix** — make the `all_reduce` body match the reduction it is partitioning: carry the reduction kind through to `sdy.all_reduce` and emit `maximum` / `minimum` / `add` accordingly, instead of hardcoding `AddOp`.
2. **Safe interim** — refuse to shard a reduction axis whose combine is not a sum: mark non-add reduction factors `kNeedReplication`. There is precedent in-tree — `RegisterCustomShardingRule.cpp:186` already pins SDPA's softmax/contraction axes that way.

Either way a guard is worth adding, since today any non-sum reduction over a sharded axis is silently wrong rather than rejected.

### Environment

tt-xla `7b44c2d24`, n300-llmbox, 8× Wormhole, SPMD mesh `(1,8)`, `CONVERT_SHLO_TO_SHARDY=1`, bf16.

**Comment by ashokkumarkannan1** [comment]
### Worked around model-side — close once the PRs land; the compiler fix is still open

- tt-xla https://github.com/tenstorrent/tt-xla/pull/6078
- tt-forge-models https://github.com/tenstorrent/tt-forge-models/pull/933

The model-side workaround pins the self-conditioning softmax output replicated (`patch_selfcond_anchor`), so the reduced axis is never sharded and the bad `all_reduce{add}` never appears. `embed_tokens` stays vocab-sharded, so the OOM fix (#6055) is untouched; the cost is a transient all-gather of [1,256,262144] bf16 = 0.125 GiB against 1.96 GiB of headroom.

Anchoring the **input** logits does not work (embeds stays 0.359807) — Shardy propagates the shard backwards, so the constraint has to sit on the tensor that feeds the sharded contraction.

**Measured:** decoder input embeds **0.359807 -> 0.999954**; step-2 logits 0.716694 -> the decoder no longer limits any case. Denoise steps **48 -> 15-18** everywhere (image 12 -> 3), and image-only went from garbage to a correct per-item image description. Peak DRAM unchanged at 9.66-10.25 GiB of 11.97.

**The underlying compiler bug is NOT fixed** and is worth keeping tracked separately: `ShardyCCLToStableHLOCCLPatterns.cpp:310-314` hardcodes an `AddOp` body for every `sdy.all_reduce`, so any **max/min** reduction over a sharded axis is silently wrong, not just softmax. Suggested fixes are in the description (carry the reduction kind through, or mark non-add reduction factors `kNeedReplication`). A guard would be worth adding either way — today this miscompiles silently rather than being rejected.
